### PR TITLE
VPN / IPsec - migrate ipsec.conf to swanctl.conf

### PR DESCRIPTION
**Pull request by AdSchellevis**
VPN/IPsec - work in progress refactoring ipsec.conf to swanctl.conf for https://github.com/opnsense/core/issues/5636

Wrap the following blocks in functions:

o generation of strongswan.conf
o write and cleanup IPsec's CA database
o write certificates used by IPsec
o write RSA Key Pairs
o construct ipsec.secrets for swanctl, combining different sources, phase1 PSK's, certificates, keypairs
o replace VPN: IPsec: Lease Status with new mvc version that uses swanctl underneath

Remap ipsec.conf to swanctl.conf attributes using https://wiki.strongswan.org/projects/strongswan/wiki/Fromipsecconf

Remove lifetime defaults from phase 1 and 2 so new connections will use Strongswan's defaults and move the option in phase above Margintime/Rekeyfuzz as these combined are being migrated (see margintime in Fromipsecconf doc).

Restructure used path directives so we can use the default structure as descried in https://docs.strongswan.org/docs/5.9/swanctl/swanctlDir.html and installed via ports.

IPsec status overview needs a small fix as well (passthrough network seems to be returned by vici, but doesn't contain a class (which crashes list_status.py).)

While here, also remove some php warnings.

Most tests performed are described in my repo https://github.com/AdSchellevis/OPNsense_IPSec_testsuite

### CODE REVIEW

**Comment by AdSchellevis** [comment]
@fichtner I think this is it for the master branch, documentation needs updates, but there's quite some time until 23.1 is released

**Comment by mimugmail** [comment]
Happy to test when in. Currently again investing time regarding EAP-TLS so I have some Labs around :)

**Comment by AdSchellevis** [comment]
@mimugmail thanks, it was quite some work refactoring this, I can certainly use an extra pair of eyes on the results.

File: src/opnsense/service/conf/actions.d/actions_ipsec.conf [hunk 1]
```diff
@@ -59,13 +65,13 @@ type:script
 message:IPsec service start
 
 [stop]
-command:/usr/local/sbin/ipsec stop
+command:/usr/local/etc/rc.d/strongswan stop
```
Comment (fichtner) [inline src/opnsense/service/conf/actions.d/actions_ipsec.conf:68]: should this not use `onestop`?

File: src/opnsense/service/conf/actions.d/actions_ipsec.conf [hunk 2]
```diff
@@ -59,13 +65,13 @@ type:script
...
 parameters:
 type:script
 message:IPsec service stop
 
 [restart]
-command:/usr/local/sbin/ipsec stop; /usr/local/sbin/pluginctl -c ipsec
```
Comment (fichtner) [inline src/opnsense/service/conf/actions.d/actions_ipsec.conf:68]: `onestop`?

File: src/etc/inc/plugins.inc.d/ipsec.inc [hunk 3]
```diff
@@ -843,820 +812,756 @@ function ipsec_configure_spd()
...
-                $ph1publickeyfile = "${publickeypath}/publickey-local-{$ph1ent['ikeid']}.pem";
-                if (!file_put_contents($ph1publickeyfile, $keyPair['publicKey'])) {
-                    log_error(sprintf('Error: Cannot write phase1 local public key file for %s', $ph1ent['name']));
-                    @unlink($ph1publickeyfile);
-                    continue;
-                }
-                @chmod($ph1publickeyfile, 0600);
+    if (!isset($ipseccfg['enable'])) {
+        /* try to stop charon */
+        mwexec('/usr/local/etc/rc.d/strongswan stop');
```
Comment (fichtner) [inline src/etc/inc/plugins.inc.d/ipsec.inc:1247]: `onestop`?

File: src/etc/inc/plugins.inc.d/ipsec.inc [hunk 4]
```diff
@@ -843,820 +812,756 @@ function ipsec_configure_spd()
...
+            echo 'Configuring IPsec VPN...';
+        }
 
-                $ph1publickeyfile = "${publickeypath}/publickey-peer-{$ph1ent['ikeid']}.pem";
-                if (!file_put_contents($ph1publickeyfile, $keyPair['publicKey'])) {
-                    log_error(sprintf('Error: Cannot write phase1 peer public key file for %s', $ph1ent['name']));
-                    @unlink($ph1publickeyfile);
-                    continue;
+        // cleanup legacy ipsec.conf files
+        if (is_file("/usr/local/etc/ipsec.conf") && is_dir("/usr/local/etc/ipsec.d")) {
```
Comment (fichtner) [inline src/etc/inc/plugins.inc.d/ipsec.inc:1267]: ipsec.conf is possibly sampled by strongswan package so always there even if we delete it (but need to double-check)

File: src/etc/inc/plugins.inc.d/ipsec.inc [hunk 5]
```diff
@@ -843,820 +812,756 @@ function ipsec_configure_spd()
...
+            foreach (new RecursiveIteratorIterator(new RecursiveDirectoryIterator("/usr/local/etc/ipsec.d")) as $x) {
+                if (is_file($x)) {
+                    @unlink($x);
                 }
-                @chmod($ph1publickeyfile, 0600);
             }
+            @unlink("/usr/local/etc/ipsec.conf");
+            @unlink("/usr/local/etc/ipsec.secrets");
```
Comment (fichtner) [inline src/etc/inc/plugins.inc.d/ipsec.inc:1274]: same for this file

File: src/etc/inc/plugins.inc.d/ipsec.inc [hunk 6]
```diff
@@ -1786,6 +1691,7 @@ function ipsec_configure_vti($verbose = false, $device = null)
         }
         if (
             empty($configured_intf[$intf])
+            || empty($intf_details['tunnel'])
```
Comment (fichtner) [inline src/etc/inc/plugins.inc.d/ipsec.inc:1694]: what's the context here?

Comment (AdSchellevis) [reply src/etc/inc/plugins.inc.d/ipsec.inc:1694]: the two lines below,  if `tunnel` doesn't exist, this `$local_configured != $intf_details['tunnel']['src_addr']` is always true, but will fail with a php error (at least as of php 8).

File: src/opnsense/service/conf/actions.d/actions_ipsec.conf (hunk 1 above)
Comment (AdSchellevis) [reply src/opnsense/service/conf/actions.d/actions_ipsec.conf]: fixed in next commit

File: src/etc/inc/plugins.inc.d/ipsec.inc (hunk 6 above)
Comment (fichtner) [reply src/etc/inc/plugins.inc.d/ipsec.inc:1690]: 👍

**Comment by fichtner** [comment]
Merged, thanks!

**Pull request merged**